sbi/mcmc/slice_numpy_vectorized.py

```diff
@@ -0,0 +1,236 @@
+# This file is part of sbi, a toolkit for simulation-based inference. sbi is licensed
+# under the Affero General Public License v3, see <https://www.gnu.org/licenses/>.
+
+import numpy as np
+import os
+import sys
+
+from tqdm import tqdm
+
+from typing import Callable, Union
+
+
+class SliceSamplerVectorized:
+    def __init__(
+        self,
+        log_prob_fn: Callable,
+        init_params: np.ndarray,
+        num_chains: int = 1,
+        tuning: int = 50,
+        verbose: bool = True,
+        init_width: Union[float, np.ndarray] = 0.01,
+        max_width: float = float("inf"),
+    ):
+        """Slice sampler in pure Numpy, vectorized evaluations across chains.
+
+        Args:
+            log_prob_fn: Log prob function.
+            init_params: Initial parameters.
+            verbose: Show/hide additional info such as progress bars.
+            tuning: Number of tuning steps for brackets.
+            init_width: Inital width of brackets.
+            max_width: Maximum width of brackets.
+        """
+        self._log_prob_fn = log_prob_fn
+
+        self.x = init_params
+        self.num_chains = num_chains
+        self.tuning = tuning
+        self.verbose = verbose
+
+        self.init_width = init_width
+        self.max_width = max_width
+
+        self.logger = open(os.devnull, "w") if not self.verbose else sys.stdout
+
+        self.n_dims = self.x.size
+
+        self._reset()
+
+    def _reset(self):
+        self.rng = np.random
+        self.state = {}
+        for c in range(self.num_chains):
+            self.state[c] = {}
+            self.state[c]["t"] = 0
+            self.state[c]["width"] = None
+            self.state[c]["x"] = None
+
+    def run(self, num_samples: int) -> np.ndarray:
+        """Runs MCMC
+
+        Args:
+            num_samples: Number of samples to generate
+
+        Returns:
+            MCMC samples
+        """
+        assert num_samples >= 0
+
+        self.n_dims = self.x.shape[1]
+
+        # Init chains
+        for c in range(self.num_chains):
+            self.state[c]["x"] = self.x[c, :]
+
+            self.state[c]["i"] = 0
+            self.state[c]["order"] = list(range(self.n_dims))
+            self.rng.shuffle(self.state[c]["order"])
+
+            self.state[c]["samples"] = np.empty([int(num_samples), int(self.n_dims)])
+
+            self.state[c]["state"] = "BEGIN"
+
+            self.state[c]["width"] = np.full(self.n_dims, self.init_width)
+
+        if self.verbose:
+            pbar = tqdm(range(self.num_chains * num_samples))
+            print("Generating MCMC samples")
+
+        num_chains_finished = 0
+        while num_chains_finished != self.num_chains:
+
+            num_chains_finished = 0
+
+            for sc in self.state.values():
```

Comment:
this collects params to evaluate across all chains, right?

Reply:
Yes, the state dict contains the whole state of all chains